ocaml/xapi/repository_helpers.ml

```diff
@@ -0,0 +1,732 @@
+    ; new_version: string
+    ; new_release: string
+    ; update_id: string option
+  }
+
+  let of_json j =
+    let open Yojson.Basic.Util in
+    {
+      name = (member "name" j |> to_string);
+      arch = (member "arch" j |> to_string);
+      new_version = (member "newVerRel" j |> member "version" |> to_string);
+      new_release = (member "newVerRel" j |> member "release" |> to_string);
+      old_version = (try Some (member "oldVerRel" j |> member "version" |> to_string)
+                     with _ -> None);
+      old_release = (try Some (member "oldVerRel" j |> member "release" |> to_string)
+                     with _ -> None);
+      update_id = (member "updateId" j |> to_string_option)
+    }
+end
+
+module Guidance = struct
+  type t =
+    | RebootHost
+    | RestartToolstack
+    | EvacuateHost
+    | RestartDeviceModel
+
+  type guidance_kind = Absolute | Recommended
+
+  let to_string = function
+    | RebootHost -> "RebootHost"
+    | RestartToolstack -> "RestartToolstack"
+    | EvacuateHost -> "EvacuateHost"
+    | RestartDeviceModel -> "RestartDeviceModel"
+
+  exception Unknown_guidance
+
+  let of_string = function
+    | "RebootHost" -> RebootHost
+    | "RestartToolstack" -> RestartToolstack
+    | "EvacuateHost" -> EvacuateHost
+    | "RestartDeviceModel" -> RestartDeviceModel
+    | _ -> raise Unknown_guidance
+
+  let of_xml = function
+    | Xml.Element ("guidance", _, [Xml.PCData v]) ->
+      of_string v
+    | _ ->
+      error "Unknown node in <absolute|recommended_guidances>";
+      raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+
+  let set_of_list l =
+    l
+    |> List.map to_string
+    |> GuidanceStrSet.of_list
+
+  let eq l s = GuidanceStrSet.equal (set_of_list l) (set_of_list s)
+
+  let eq_set1 = eq [EvacuateHost; RestartToolstack]
+
+  let eq_set2 = eq [RestartDeviceModel; RestartToolstack]
+
+  let eq_set3 = eq [RestartDeviceModel; EvacuateHost]
+
+  let eq_set4 = eq [EvacuateHost; RestartToolstack; RestartDeviceModel]
+
+  let assert_valid_guidances = function
+    | [RebootHost]
+    | [EvacuateHost]
+    | [RestartToolstack]
+    | [RestartDeviceModel] -> ()
+    | l when eq_set1 l ->
+      (* EvacuateHost and RestartToolstack *)
+      ()
+    | l when eq_set2 l ->
+      (* RestartDeviceModel and RestartToolstack *)
+      ()
+    | l when eq_set3 l ->
+      (* RestartDeviceModel and EvacuateHost *)
+      ()
+    | l when eq_set4 l ->
+      (* EvacuateHost, RestartToolstack and RestartDeviceModel *)
+      ()
+    | l ->
+      let msg =
+        Printf.sprintf "Found wrong guidance(s) before applying updates: %s"
+          (String.concat ";" (List.map to_string l))
+      in
+      raise Api_errors.(Server_error (internal_error, [msg]))
+
+  let resort_guidances gs=
+    match GuidanceStrSet.find_opt "RebootHost" gs with
+    | Some _ -> GuidanceStrSet.singleton "RebootHost"
+    | None -> gs
+end
+
+module Applicability = struct
+  type inequality = Lt | Eq | Gt
```

Comment:
A better name would be `order`

Reply:
IIRC this model the version constraints, which is subtly different from ordering. This means that `Lte` and `Gte` need to be modelled. This is done currently with a list like `[Lt; Eq]`.
I would prefer if Lte and Gte were added to the variant type instead of handling lists.

Reply:
`Lte` and `Gte` are actually useless in actual comparison. For example, `compare s1 s2 ` will return `Lt`, `Eq`, or `Gt`. It will not return `Lte` or `Gte`.

Reply:
The use of lists (which represent conjunction) also seems ad-hoc to me. We are parsing a string "<=" as a version constrained and need to check whether two versions respect the constrained. It is more important to model the problem than what `compare` returns. Implementing the `<=` constraint it terms of compare is easy: `leq x y = lt x Y || eq x y`. I don't see why we need to represent what `compare` returns as a sum type.

Reply:
Thanks. In this case, I think following should present the modeling:
`
type inequality = Lt | Eq | Gt | Lte | Gte

type order = Lt_ | Eq_ | Gt_

let lt (v1, r1) (v2, r2) ->
  match (compare v1 v2), (compare r1 r2) with
  | Lt_, _ | Eq_, Lt_ -> true
  | _ -> false

let gt (v1, r1) (v2, r2) ->
  match (compare v1 v2), (compare r1 r2) with
  | Gt_, _ | Eq_, Gt_ -> true
  | _ -> false

let eq (v1, r1) (v2, r2) ->
  match (compare v1 v2), (compare r1 r2) with
  | Eq_, Eq_ -> true
  | _ -> false

let lte vr1 vr2 = lt vr1 vr2 || eq vr1 vr2
let gte vr1 vr2 = lt vr1 vr2 || eq vr1 vr2

let eval_applicability vr1 vr2
match inequality
| Lt -> lt vr1 vr2
| Lte -> lte vr1 vr2
| Gt  -> gt vr1 vr2
| Gte -> gte vr1 vr2
| Eq -> eq vr1 vr2
`

Reply:
I think it would be better to avoid the `Invalid` variant - so that if you have a value of type `inequality`, you don't have to do an awkward match on `Invalid` (unless of course `Invalid` is genuinely a state we expect to see)

Reply:
I don't know if changing these types makes thing better. It certainly doesn't look _simpler_ to me.

Reply:
I think this is a log clearer. The underscores could be avoided by using `LT`, `GT`, `EQ` for `order`.

Reply:
> I think it would be better to avoid the Invalid variant - so that if you have a value of type inequality, you don't have to do an awkward match on Invalid (unless of course Invalid is genuinely a state we expect to see)

A default value is needed for this. Not found a better way :(

Reply:
> An alternative name for for inequality could be constraint.

For the name, this is inherited from the tag name of updateinfo.xml.  I think it would be better to keep the consistency.

Reply:
> A default value is needed for this. Not found a better way :(
You could use `inequality Option` instead. I think the same should be done with the rest of the fields in the records.
I see that's a pain to contruct because of how xml is parsed. Maybe you can use a "partial record" with optional fields, at the end of the xml parsing you can discard records with None fields and construct a "complete" record if you're successful.

Reply:
> A default value is needed for this. Not found a better way :(

you can use `None` for default values (from the `option` type). you should be able to write a parse function with the following signature:

```
parse_xml : string -> t option
```
where `t` represents a completely valid result (i.e. the parse function should handle invalid xml states for us, so we don't have worry about it elsewhere in the code)